socorro/external/es/crashstorage.py

```diff
@@ -264,32 +272,164 @@ def build_document(src, crash_document, fields, all_keys):
+        shards_per_index=10,
+    ):
+        super().__init__()
 
-        self.es_context = self.config.elasticsearch.elasticsearch_class(
-            config=self.config.elasticsearch
-        )
-        self.metrics = markus.get_metrics(namespace)
+        self.client = self.build_client(url=url, timeout=timeout)
+
+        self.metrics = markus.get_metrics(metrics_prefix)
+
+        self.index = index
+        self.index_regex = index_regex
+        self.retention_policy = retention_policy
+        self.doctype = doctype
+        self.shards_per_index = shards_per_index
 
         # Cached answers for things that don't change
         self._keys_for_indexable_fields_cache = None
         self._keys_for_mapping_cache = {}
+        self._mapping_cache = {}
+
+    @classmethod
+    def build_client(cls, url, timeout):
+        return ConnectionContext(url=url, timeout=timeout)
+
+    def get_index_template(self):
+        """Return template for index names."""
+        return self.index
+
+    def get_index_for_date(self, date):
+        """Return the index name for a given date.
+
+        Index names are generated by filling in the date parameters for the
+        specified date in the index template.
+
+        :arg datetime date: the datetime to use to generate the index template
+
+        :returns: index name as a str
+
+        """
+        template = self.get_index_template()
+        return date.strftime(template)
+
+    def get_doctype(self):
+        """Return doctype."""
+        return self.doctype
+
+    def get_retention_policy(self):
+        """Return retention policy in weeks."""
+        return self.retention_policy
+
+    def get_socorro_index_settings(self, mappings):
+        """Return a dictionary containing settings for an Elasticsearch index."""
+        return {
+            "settings": {
+                "index": {
+                    "number_of_shards": self.shards_per_index,
+                    "query": ES_QUERY_SETTINGS,
+                    "analysis": ES_CUSTOM_ANALYZERS,
+                }
+            },
+            "mappings": mappings,
+        }
+
+    def get_mapping(self, index_name, es_doctype, reraise=False):
+        """Retrieves the mapping for a given index and doctype
+
+        NOTE(willkg): Mappings are cached on the ConnectionContext instance. If you
+        change the indices (like in tests), you should get a new ConnectionContext
+        instance.
+
+        :arg str index_name: the index to retrieve the mapping for
+        :arg str es_doctype: the doctype to retrieve the mapping for
+        :arg bool reraise: True if you want this to reraise a NotFoundError; False
+            otherwise
+
+        :returns: mapping as a dict or None
+
+        """
+        cache_key = f"{index_name}::{es_doctype}"
+        mapping = self._mapping_cache.get(cache_key)
+        if mapping is None:
+            try:
+                resp = self.client.indices_client().get_mapping(index=index_name)
```

Comment:
The crashstorage code shouldn't access the indices_client directly. We should have an intermediary method.